### ISSUE TITLE
Connecting multiple raspberry pi to same universe

### ISSUE THREAD

**Issue by nagyation** [OpenLightingProject/ola#1565]
What it's needed to be able to connect more than one raspberry pi together, to make one master and others slave, is there an implementation in the library for that?

**Comment by peternewman** [comment]
Hi @nagyation ,

All you need to do is choose a protocol (I'd recommend E1.31 or Art-Net) and then patch one Pi as an output and others as input to the same universe number and protocol type. It should be as simple as that.

**Comment by nagyation** [comment]
I want to do this from the library dynamicly, since we need to control who is the master on demand...

**Comment by nagyation** [comment]
It's actually a C project we use the Ola library in it to control rgb led over dmx

**Comment by peternewman** [comment]
Can you give a little bit more detail? Would this assumption be correct?

3 Pis, each with some LEDs attached
Pi A is in control of itself and all other LED colours
Something happens
Pi B is in control of itself and all other LED colours

Or something else entirely?

**Comment by nagyation** [comment]
Pi A is master gets rgb data and sends it to all other including itself
And using some conditions, on demand we can make B is master and all other recieve rgb data from it

**Comment by peternewman** [comment]
> Pi A is master gets rgb data and sends it to all other including itself
> And using some conditions, on demand we can make B is master and all other recieve rgb data from it

I think you should be able to do this with a bit of patching using the API to patch/unpatch. Or possibly just to change the priority of the various devices generating data.

How is the DMX output leaving each Pi to control the LEDs, or is it via SPI or something?

**Comment by nagyation** [comment]
Thanks for your replies, But how actually the connection between them work? They are connected to the same network or something?
And where to search about that in the api? 
And yes it's connected to it using spi

**Comment by nagyation** [comment]
What really bother me is.. How do universe get connected to each other? And how they communicate? Do I need to do some configuration to make one of the raspberry to host the universe? And the others must know its ip and connect to it?
I don't really understand the api, and by sending to the channels of the universe, does it broadcast directly to the dmx then to the rgb?
Or I need to recieve the data at each raspberry and put it on the rgb channels?

I thought to do my server hosts code, but if it's already implemented it will be easier..

**Comment by peternewman** [comment]
If they're all on the same Ethernet or WiFi network it should just work (E1.31 uses multicast, Art-Net uses broadcast and discovery). The device that's patched as an input port will generate data, and anything else patched to the same universe as an output will receive the data.

If I were you I'd start experimenting using either the CLI tools or the web UI, they're essentially just examples for the API. When you've got it doing what you want via that we can assist in helping you to write some equivalent code.

Also let us know your language preference, there are mature C++ and Python APIs, and a reasonable Java one and some Go attempts.

**Comment by nagyation** [comment]
First thanks for this fast support,
Okay the problem is that I'm working remotely on this part I don't have direct access to the hardware currently, but they managed to configure it using arnet over the web ui, so is there any api or commands I could use? and where to search, the proposed method is that the user at will, will choose the master using some kind of interface.. so I don't know how the data is sent and received, but if also there is a place to add some kind of message command, to broadcast the new master, thus the old master patch itself to slave..

**Comment by peternewman** [comment]
Our JSON/Web API is the only one you can use remotely, so would be the best place to start:
https://wiki.openlighting.org/index.php/OLA_JSON_API

A screenshot or the output of ola_patch on each machine would help to clarify things too.

Does the source of data HAVE to be local to each Pi? Does each Pi HAVE to only output DMX for some of the RGB LEDs? What is the actual data source into each Pi? Are they reacting to something? What? How?

> It's actually a C project we use the Ola library in it to control rgb led over dmx

I assume you mean you're acting as an OLA client into each local OLA instance?

**Comment by nagyation** [comment]
I meant that I code for them remotely, I don't have direct access to project, so it's a little hard to debug, that's why I wanted to understand the arnet thing so I could write a code that does the job..

**Comment by nagyation** [comment]
Yes, each of raspberry has its own source of data, and interact with user using gui, they are connected over network, and yes we use ola client api for outputing on dmx channels

**Comment by nagyation** [comment]
The steps taken with the Web interface is just, making the raspberry Direction of data Output with arnet port, and all other input, with the same universe..
So I think all i have to do is register to the same universe, and that's simple using client api..
and the other trick is to open connection using arnet.. I don't know if it's already supported in c++, if not, I think using cli is a workaround ? https://opendmx.net/index.php/OLA_Command_Line_Tools

**Comment by Keeper-of-the-Keys** [comment]
As far as I understood what you were trying to accomplish is a hot standby with 2 or more OLA boxes and you want to know what commands/API endpoints in OLA will allow you to convert a box standing by from standby to master?

First of all the actual decision to convert would need to be based on heartbeat or something similar as far as I know we don't have that built into OLA, may be cool to try and do in the future though.

**Comment by nagyation** [comment]
What you say is that there isn't a way of setting device as master? or the concept of hot switch itself?

> As far as I understood what you were trying to accomplish is a hot standby with 2 or more OLA boxes and you want to know what commands/API endpoints in OLA will allow you to convert a box standing by from standby to master?
> 
> First of all the actual decision to convert would need to be based on heartbeat or something similar as far as I know we don't have that built into OLA, may be cool to try and do in the future though.

**Comment by Keeper-of-the-Keys** [comment]
I'm saying that to the best of my knowledge we do not have anything built into OLA that allows one box to detect the health of another box and thus decide whether or not it should take over, of course I could be wrong.

Thus whatever decides which device is distributing the data is external to OLA, it could be some form of a heartbeat deamon with "shoot the other node in the head" capability or some other paradigm.

As @peternewman already stated to decide what drives a univers you just need to patch is as input or as output for which an API exists.

As far as "hot standby" like functionality goes some IP based protocols may have "master"/"slave" functionality (I am not familiar enough with them to tell you for certain), the physical/"last mile" interfaces (DMX, SPI, TTL and the various other LED control protocols) do not support such a model.
With DMX the way this is usually accomplished is a device that has 2 inputs and 1 output driving the line which will switch either manually or when DMX failure is detected on the main line.

After re-reading the thread I also feel that I have not managed to sufficiently understand what you are trying to accomplish making helping you harder.

**Comment by nagyation** [comment]
Okay so what I currently want, is that how to patch a device as a master or Art-Net.. , using library or command line, I can't find any documentation/tutorial on that..
And by seeing the current approach I might find a solution..

**Comment by Keeper-of-the-Keys** [comment]
The ola_patch command is porbably what you are looking for on CLI.

**Comment by nagyation** [comment]
Thanks for help, I will post all my experience here after I finish

**Comment by nagyation** [comment]
Okay so I enabled artnet using the config files and enabled always broadcast, but for make it work each time  I have to reload plug-ins from the web interface?
I need it to load it in boot so I can directly patch the ports..

**Comment by peternewman** [comment]
> Okay so what I currently want, is that how to patch a device as a master or Art-Net.. , using library or command line, I can't find any documentation/tutorial on that..

> The ola_patch command is probably what you are looking for on CLI.

As @Keeper-of-the-Keys says the ola_patch command is the CLI tool, it just calls the client API tools, documentation here:

http://docs.openlighting.org/ola/doc/latest/classola_1_1client_1_1_ola_client.html

Specifically you'll want patch:

http://docs.openlighting.org/ola/doc/latest/classola_1_1client_1_1_ola_client.html#ab2e8198e0e2ea81ec8dfff8c4c28b6e4

**Comment by peternewman** [comment]
> Okay so I enabled artnet using the config files and enabled always broadcast, but in order to make it work each time I have to reload plug-ins from the web interface?
> I need it to load it in boot so I can directly patch the ports..

@nagyation the issue is because OLAd is starting before the network interface is up.

There is lots of discussion in the mailing list:
https://groups.google.com/forum/#!forum/open-lighting

Such as here:
https://groups.google.com/forum/#!topic/open-lighting/zDGJTCynLwc

One possible workaround is here:
https://groups.google.com/forum/#!topic/open-lighting/kUj4gKZ1_R8
